File: app/src/main/java/su/linka/pictures/activity/MainActivity.java
package su.linka.pictures.activity;

import static su.linka.pictures.components.SetContextDialog.DELETE;
import static su.linka.pictures.components.SetContextDialog.EDIT;
import static su.linka.pictures.components.SetContextDialog.OPEN;
import static su.linka.pictures.components.SetContextDialog.RENAME;

import androidx.activity.result.ActivityResult;
import androidx.activity.result.ActivityResultCallback;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.Toast;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.firebase.analytics.FirebaseAnalytics;

import java.io.IOException;

import su.linka.pictures.Callback;
import su.linka.pictures.R;
import su.linka.pictures.SetManifest;
import su.linka.pictures.SetsManager;
import su.linka.pictures.components.ConfirmDialog;
import su.linka.pictures.components.InputDialog;
import su.linka.pictures.components.ParentPasswordDialog;
import su.linka.pictures.components.SetContextDialog;

public class MainActivity extends AppCompatActivity {

    private ListView setsList;
    private ArrayAdapter<SetManifest> adapter;
    private FirebaseAnalytics mFirebaseAnalytics;
    private SetsManager setsManager;
    private ActivityResultLauncher<Intent> activityLauncher ;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        mFirebaseAnalytics = FirebaseAnalytics.getInstance(this);

        setsManager = new SetsManager(this);
        activityLauncher = registerForActivityResult(
                new ActivityResultContracts.StartActivityForResult(),
                new ActivityResultCallback<ActivityResult>() {
                    @Override
                    public void onActivityResult(ActivityResult result) {
                        if (result.getResultCode() == Activity.RESULT_OK) {
                            loadSetsList();
                        }
                    }
                });

        loadDefaultSets();

         setsList = findViewById(R.id.sets_list);

        adapter = new ArrayAdapter<SetManifest>(this, androidx.appcompat.R.layout.support_simple_spinner_dropdown_item);
        setsList.setAdapter(adapter);
        loadSetsList();
        setsList.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                SetManifest manifest = adapter.getItem(position);
                openSet( manifest);

            }
        });

        setsList.setOnItemLongClickListener(new AdapterView.OnItemLongClickListener() {
            @Override
            public boolean onItemLongClick(AdapterView<?> parent, View view, int position, long id) {
                SetManifest manifest = adapter.getItem(position);
                SetContextDialog
                        .show(view.getContext(), manifest.toString(), new Callback<Integer>() {
                            @Override
                            public void onDone(Integer result) {
                                switch (result){
                                    case OPEN:
                                        openSet(manifest);
                                        break;
                                    case EDIT:
                                        editSet(manifest);
                                        break;
                                    case RENAME:
                                        renameSet(manifest);
                                        break;
                                    case DELETE:
                                        deleteSet(manifest);
                                        break;
                                }
                            }

                            @Override
                            public void onFail(Exception error) {

                            }
                        });
                return true;
            }
        });

        FloatingActionButton fab = findViewById(R.id.fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                createSet();
            }
        });
    }

    private void renameSet(SetManifest manifest) {
        InputDialog
                .showDialog(this, R.string.rename, new Callback<String>() {
                    @Override
                    public void onDone(String result) {
                        setsManager.rename(manifest, result);
                        loadSetsList();
                    }

                    @Override
                    public void onFail(Exception error) {

                    }
                });

    }

    private void deleteSet(SetManifest manifest) {
        ConfirmDialog
                .showConfirmDialog(this, R.string.delete, new Callback() {
                    @Override
                    public void onDone(Object o) {
                        setsManager.delete(manifest);
                        loadSetsList();
                    }

                    @Override
                    public void onFail(Exception error) {

                    }
                });
    }

    private void openSet(SetManifest manifest) {
        Intent intent = new Intent(this, GridActivity.class);
        Bundle b = new Bundle();
        b.putString("file", manifest.toString()); //Your id
        intent.putExtras(b);
        startActivity(intent);
    }
    private void editSet(SetManifest manifest) {
        Intent intent = new Intent(this, SetEditActivity.class);
        Bundle b = new Bundle();
        b.putString("file", manifest.toString()); //Your id
        intent.putExtras(b);
        startActivity(intent);
    }

    private void createSet() {
        Context context = this;
        ParentPasswordDialog
                .showDialog(this, new Callback() {
                    @Override
                    public void onDone(Object o) {
                        Intent intent = new Intent(context, SetEditActivity.class);
                        activityLauncher.launch(intent);
                    }
                    @Override
                    public void onFail(Exception error) {

                    }
                });
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {

        getMenuInflater().inflate(R.menu.main_activity_menu, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if(item.getItemId()==R.id.settings){
            Context context = this;
            ParentPasswordDialog.showDialog(getWindow().getContext(), new Callback() {
                        @Override
                        public void onDone(Object o) {

                    Intent intent = new Intent(context, SettingsActivity.class);
                    startActivity(intent);
                }

                @Override
                public void onFail(Exception error) {

                }
            });
        }
        return super.onOptionsItemSelected(item);
    }

    private void loadSetsList() {

        SetManifest[] sets = setsManager
                .getSets();
        adapter.clear();
        adapter.addAll(sets);
        adapter.notifyDataSetChanged();
    }

    protected void loadDefaultSets(){


        try {
            setsManager.loadDefaultSets();
        } catch (IOException e) {

            Log.e(getClass().getCanonicalName(), "loadSets: ", e);
            Toast.makeText(this, "copy assets error", Toast.LENGTH_LONG).show();

        }
    }

}

File: app/src/main/java/su/linka/pictures/activity/MainActivity.kt
package su.linka.pictures.activity

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView
import android.widget.GridView
import android.widget.Toast
import androidx.activity.result.ActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import com.google.android.material.appbar.MaterialToolbar
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.firebase.analytics.FirebaseAnalytics
import su.linka.pictures.AnalyticsEvents
import su.linka.pictures.Callback
import su.linka.pictures.R
import su.linka.pictures.SetManifest
import su.linka.pictures.SetsAdapter
import su.linka.pictures.SetsManager
import su.linka.pictures.components.InputDialog
import su.linka.pictures.components.ParentPasswordDialog
import su.linka.pictures.components.SetContextDialog
import java.io.IOException

class MainActivity : AppCompatActivity() {

    private lateinit var setsList: GridView
    private lateinit var adapter: SetsAdapter
    private lateinit var analytics: FirebaseAnalytics
    private lateinit var setsManager: SetsManager

    private val activityLauncher = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) { result: ActivityResult ->
        if (result.resultCode == Activity.RESULT_OK) {
            loadSetsList()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val toolbar: MaterialToolbar = findViewById(R.id.toolbar)
        setSupportActionBar(toolbar)

        analytics = FirebaseAnalytics.getInstance(this)
        supportActionBar?.setTitle(R.string.your_sets)

        setsManager = SetsManager(this)
        adapter = SetsAdapter(this)

        setsList = findViewById(R.id.sets_list)
        setsList.adapter = adapter

        loadDefaultSets()
        loadSetsList()

        setsList.onItemClickListener = AdapterView.OnItemClickListener { _, _, position, _ ->
            adapter.getItem(position)?.let { manifest ->
                openSet(manifest)
            }
        }

        setsList.onItemLongClickListener = AdapterView.OnItemLongClickListener { view, _, position, _ ->
            val manifest = adapter.getItem(position) ?: return@OnItemLongClickListener true
            SetContextDialog.show(view.context, manifest.name, object : Callback<Int>() {
                override fun onDone(result: Int) {
                    when (result) {
                        SetContextDialog.OPEN -> openSet(manifest)
                        SetContextDialog.EDIT -> editSet(manifest)
                        SetContextDialog.RENAME -> renameSet(manifest)
                        SetContextDialog.DELETE -> deleteSet(manifest)
                    }
                }

                override fun onFail(error: Exception?) = Unit
            })
            true
        }

        findViewById<FloatingActionButton>(R.id.fab).setOnClickListener {
            createSet()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main_activity_menu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return if (item.itemId == R.id.settings) {
            ParentPasswordDialog.showDialog(this, object : Callback<Any?>() {
                override fun onDone(result: Any?) {
                    analytics.logEvent(AnalyticsEvents.OPEN_SETTINGS, null)
                    startActivity(Intent(this@MainActivity, SettingsActivity::class.java))
                }

                override fun onFail(error: Exception?) = Unit
            })
            true
        } else {
            super.onOptionsItemSelected(item)
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        adapter.shutdown()
    }

    private fun renameSet(manifest: SetManifest) {
        InputDialog.showDialog(this, R.string.rename, object : Callback<String>() {
            override fun onDone(result: String) {
                setsManager.rename(manifest, result)
                loadSetsList()
            }

            override fun onFail(error: Exception?) = Unit
        })
    }

    private fun deleteSet(manifest: SetManifest) {
        su.linka.pictures.components.ConfirmDialog.showConfirmDialog(this, R.string.delete, object : Callback<Any?>() {
            override fun onDone(result: Any?) {
                setsManager.delete(manifest)
                loadSetsList()
            }

            override fun onFail(error: Exception?) = Unit
        })
    }

    private fun openSet(manifest: SetManifest) {
        analytics.logEvent(AnalyticsEvents.OPEN_SET, null)
        val intent = Intent(this, GridActivity::class.java).apply {
            putExtra(GridActivity.EXTRA_FILE, manifest.name)
        }
        startActivity(intent)
    }

    private fun editSet(manifest: SetManifest) {
        analytics.logEvent(AnalyticsEvents.EDIT_SET, null)
        val intent = Intent(this, SetEditActivity::class.java).apply {
            putExtra(SetEditActivity.EXTRA_FILE, manifest.name)
        }
        startActivity(intent)
    }

    private fun createSet() {
        analytics.logEvent(AnalyticsEvents.CREATE_SET, null)
        ParentPasswordDialog.showDialog(this, object : Callback<Any?>() {
            override fun onDone(result: Any?) {
                val intent = Intent(this@MainActivity, SetEditActivity::class.java)
                activityLauncher.launch(intent)
            }

            override fun onFail(error: Exception?) = Unit
        })
    }

    private fun loadSetsList() {
        val sets = setsManager.getSets()
        adapter.clear()
        adapter.addAll(sets.toList())
        adapter.notifyDataSetChanged()
    }

    private fun loadDefaultSets() {
        try {
            setsManager.loadDefaultSets()
        } catch (error: IOException) {
            error.printStackTrace()
            Toast.makeText(this, R.string.copy_assets_error, Toast.LENGTH_LONG).show()
        }
    }

}
